src/next-collection-view.js

```diff
@@ -0,0 +1,615 @@
+      return;
+    }
+
+    this._destroyEmptyView();
+
+    this.triggerMethod('before:sort', this);
+
+    this._sortChildren();
+
+    const filteredViews = this._filterChildren(removedViews);
+
+    this._detachChildren(filteredViews.detaching);
+
+    // Check if all views have been filtered out
+    if (this.isEmpty(!filteredViews.attaching.length)) {
+      this._showEmptyView()
+    } else {
+      this._renderChildren(filteredViews.attaching);
+    }
+
+    // Sort event reflects the DOM sort and not the children sort
+    this.triggerMethod('sort', this);
+  },
+
+  isEmpty(allViewsFiltered) {
+    return allViewsFiltered || !this.children.length;
+  },
+
+  _showEmptyView() {
+    const EmptyView = this._getEmptyView();
+
+    if (!EmptyView) {
+      return;
+    }
+
+    const options = this._getEmptyViewOptions();
+
+    this.emptyRegion.show(new EmptyView(options));
+  },
+
+  // Retrieve the empty view class
+  _getEmptyView() {
+    const emptyView = this.emptyView;
+
+    if (!emptyView) { return; }
+
+    return this._getView(emptyView);
+  },
+
+  // Remove the emptyView
+  _destroyEmptyView() {
+
+    // Only empty if a view is show so the region
+    // doesn't detach any other unrelated HTML
+    if (this.emptyRegion.hasView()) {
+      this.emptyRegion.empty();
+    }
+  },
+
+  //
+  _getEmptyViewOptions() {
+    const emptyViewOptions = this.emptyViewOptions || this.childViewOptions;
+
+    if (_.isFunction(emptyViewOptions)) {
+      return emptyViewOptions.call(this);
+    }
+
+    return emptyViewOptions;
+  },
+
+  // Sorts views by viewComparator and sets the children to the new order
+  _sortChildren() {
+    const viewComparator = this.getViewComparator();
+
+    const sortedViews = this._sortChildrenBy(viewComparator);
+
+    this.children._set(sortedViews);
+  },
+
+  // If viewComparator is overriden it will be returned here.
+  // Additionally override this function to provide custom
+  // viewComparator logic
+  getViewComparator() {
+    return this.viewComparator || this._viewComparator;
+  },
+
+  // Default internal view comparator that order the views by
+  // the order of the collection
+  _viewComparator(view) {
+    return this.collection.indexOf(view.model);
+  },
+
+  _sortChildrenBy(comparator) {
+    const views = this.children._views;
+
+    if (typeof comparator === 'string') {
+      return _.sortBy(views, (view) => {
+        return view.model.get(comparator);
+      });
+    }
+
+    comparator = _.bind(comparator, this);
+
+    if (comparator.length === 1) {
+      return _.sortBy(views, comparator);
+    }
+
+    return views.sort(comparator);
+  },
+
+  setFilter(filter, {preventRender} = {}) {
+    const canBeRendered = this._isRendered && !this._isDestroyed;
+    const filterChanged = this.filter !== filter;
+    const shouldRender = canBeRendered && filterChanged && !preventRender;
+
+    this.filter = filter;
+
+    if (shouldRender) {
+      this._showChildren();
```

Comment:
`this._showChildren` will sort `this.children._views` again when it really needs to do line
https://github.com/marionettejs/backbone.marionette/pull/3244/files#diff-c4fcc4edadcafa174aa6de6124ecbeacR294 through https://github.com/marionettejs/backbone.marionette/pull/3244/files#diff-c4fcc4edadcafa174aa6de6124ecbeacR303